core/src/avm2/verify.rs

```diff
@@ -0,0 +1,1126 @@
+                        .unwrap()
+                        .ok_or_else(|| {
+                            Error::AvmError(
+                                verify_error(
+                                    activation,
+                                    &format!(
+                                        "Error #1014: Class {} could not be found.",
+                                        multiname.to_qualified_name(activation.context.gc_context)
+                                    ),
+                                    1014,
+                                )
+                                .expect("Error should construct"),
+                            )
+                        })?;
+                }
+
+                _ => {}
+            }
+
+            i += 1;
+        }
+
+        if !save_current_work {
+            worklist.pop();
+        }
+    }
+
+    Ok(())
+}
+
+fn optimize<'gc>(
+    activation: &mut Activation<'_, 'gc>,
+    method: &BytecodeMethod<'gc>,
+    code: &mut Vec<Op>,
+    jump_targets: HashSet<i32>,
+) {
+    // These make the code less readable
+    #![allow(clippy::manual_filter)]
+    #![allow(clippy::single_match)]
+
+    // This can probably be done better by recording the receiver in `Activation`,
+    // but this works since it's guaranteed to be set in `Activation::from_method`.
+    let this_value = activation.local_register(0);
+
+    let this_class = if let Some(this_class) = activation.subclass_object() {
+        if this_value.is_of_type(activation, this_class.inner_class_definition()) {
+            Some(this_class)
+        } else {
+            None
+        }
+    } else {
+        None
+    };
+
+    // Initial set of local types
+    let mut local_types = vec![None; method.body().unwrap().num_locals as usize];
+    local_types[0] = this_class;
+
+    // Invalidate local types if they should be invalidated
+    for op in &*code {
+        match op {
+            Op::SetLocal { index }
+            | Op::Kill { index }
+            | Op::IncLocal { index }
+            | Op::IncLocalI { index }
+            | Op::DecLocal { index }
+            | Op::DecLocalI { index } => {
+                if (*index as usize) < local_types.len() {
+                    local_types[*index as usize] = None;
+                }
+            }
+            Op::HasNext2 {
+                object_register,
+                index_register,
+            } => {
+                if (*object_register as usize) < local_types.len() {
+                    local_types[*object_register as usize] = None;
+                }
+                if (*index_register as usize) < local_types.len() {
+                    local_types[*index_register as usize] = None;
+                }
+            }
+            _ => {}
+        }
+    }
+
+    let mut previous_op = None;
+    for (i, op) in code.iter_mut().enumerate() {
+        if let Some(previous_op_some) = previous_op {
```

Comment:
Can we unnest this a bit? (And definitely rename `previous_op_some`)

e.g. 
```rust
let previous_op = if let Some(previous_op) = previous_op {
  previous_op
} else {
  previous_op = Some(op.clone());
}
```

In theory we could even do something like
```rust
previous_op = code.first();
for (...) in code.iter_mut().enumerate().skip(1) {
```

Reply:
I can't do `let Some(previous_op) = previous_op` because the optimizer needs to set the `Option` value of `previous_op`. I'll try unnesting it.

Reply:
I tried unnesting it but I realized that there are likely going to be optimizations that we can do that would work despite `previous_op` being `None`, and/or those that would work even when the next op is on a jump target.